### PR TITLE
feat: teach LayoutBatchStream to filter by indices

### PR DESCRIPTION
**Pull request by danking**
I have some thoughts on how to make the stream.rs logic easier to follow; however, this, I think, is the fewest changes to achieve indices filtering.

This change evaluates the filter on any row in a chunk containing at least one kept index. I have some thoughts on how to avoid evaluating the filter at a sub-chunk level but they require more substantial changes.

### CODE REVIEW

File: vortex-serde/src/layouts/read/builder.rs [hunk 1]
```diff
@@ -87,22 +88,28 @@ impl<R: VortexReadAt> LayoutBatchStreamBuilder<R> {
...
+            .map(|row_filter| {
                 footer.layout(
-                    Scan::new(Some(Arc::new(filter.clone()))),
+                    Scan::new(Some(Arc::new(row_filter))),
                     RelativeLayoutCache::new(message_cache.clone(), footer_dtype),
                 )
             })
             .transpose()?;
 
+        let indices_mask = self
```
Comment (robert3005) [inline vortex-serde/src/layouts/read/builder.rs]: you could build the mask in the `with` method and then we can allow boolean and integer masks?

File: vortex-serde/src/layouts/tests.rs [hunk 2]
```diff
@@ -551,3 +551,252 @@ async fn filter_and() {
...
+    //         .build()
+    //         .await
+    //         .unwrap()
+    //         .read_all()
+    //         .await
+    //         .unwrap()
+    //         .into_struct()
+    //         .unwrap();
+
+    // assert_eq!(actual_kept_array.len(), 0);
```
Comment (robert3005) [inline vortex-serde/src/layouts/tests.rs]: keep/remove?

File: vortex-serde/src/layouts/read/stream.rs [hunk 3]
```diff
@@ -116,21 +120,38 @@ impl<R: VortexReadAt + Unpin + 'static> Stream for LayoutBatchStream<R> {
...
+                            }
+                        }
+
+                        self.current_selector = Some(new_selector);
                         StreamingState::Filter
                     } else {
+                        if let Some(indices_mask) = &self.indices_mask {
+                            new_selector.and_inplace(
+                                &indices_mask.slice(new_selector.begin(), new_selector.end()),
+                            )?;
```
Comment (robert3005) [inline vortex-serde/src/layouts/read/stream.rs]: we still want this to happen in filter case. We should handle non dense selection in filter

Comment (robert3005) [reply vortex-serde/src/layouts/read/stream.rs]: hmmmm, ok let's file this as an issue. I think this could be a followup

**Review by robert3005 — APPROVED** [review]

File: vortex-serde/src/layouts/read/stream.rs (hunk 3 above)
Comment (robert3005) [reply vortex-serde/src/layouts/read/stream.rs]: filed #1243

File: vortex-serde/src/layouts/read/builder.rs (hunk 1 above)
Comment (danking) [reply vortex-serde/src/layouts/read/builder.rs]: Sure, I'll do that

**Comment by danking** [comment]
All comments addressed.

**Comment by robert3005** [comment]
@danking with this pr could you switch https://github.com/spiraldb/vortex/blob/develop/bench-vortex/benches/random_access.rs to use it? This would be a real benchmark now instead of wishful thinking

**Comment by danking** [comment]
Yes

**Comment by danking** [comment]
Following up with bench change.

**Pull request merged**